File: 622-design-circular-queue/622-design-circular-queue.cpp
struct Node{
    int val;
    Node* nex;
    Node* pre;
    Node(int v):val(v){};
};
class MyCircularQueue {
public:
    int c;
    int len;
    Node* head;
    MyCircularQueue(int k) {
        c=k;
        len=0;
        head=new Node(0);
        head->nex=head;
        head->pre=head;
    }
    
    bool enQueue(int value) {
        if(len==c)return false;
        len++;
        Node* root= new Node(value);
        Node* pre = head->pre;
        pre->nex=root;
        root->nex=head;
        
        head->pre=root;
        root->pre=pre;
        return true;
    }
    
    bool deQueue() {
        if (len==0)return false;
        Node* nex= head->nex->nex;
        head->nex=nex;
        nex->pre=head;
        len--;
        return true;
    }
    
    int Front() {
        return len==0?-1:head->nex->val;
    }
    
    int Rear() {
        return len==0?-1:head->pre->val;
    }
    
    bool isEmpty() {
        return len==0;
    }
    
    bool isFull() {
        return len==c;
    }
};

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * MyCircularQueue* obj = new MyCircularQueue(k);
 * bool param_1 = obj->enQueue(value);
 * bool param_2 = obj->deQueue();
 * int param_3 = obj->Front();
 * int param_4 = obj->Rear();
 * bool param_5 = obj->isEmpty();
 * bool param_6 = obj->isFull();
 */

File: 622-design-circular-queue/622-design-circular-queue.py
class Node:
    def __init__(self,val=0):
        self.val=val
        self.nex=None
        self.pre=None
        
class MyCircularQueue:

    def __init__(self, k: int):
        self.len=0
        self.k=k
        self.head=Node()
        self.head.nex=self.head.pre=self.head

    def enQueue(self, value: int) -> bool:
        if self.len==self.k:
            return False
        pre=self.head.pre
        root=Node(value)
        
        pre.nex=root
        root.nex=self.head
        
        self.head.pre=root
        root.pre=self.head
        self.len+=1
        return True

    def deQueue(self) -> bool:
        if self.len==0:
            return False
        nex=self.head.nex.nex
        self.head.nex=nex
        nex.pre=self.head
        self.len-=1
        return True

    def Front(self) -> int:
        if self.len==0:
            return -1
        return self.head.nex.val

    def Rear(self) -> int:
        if self.len==0:
            return -1
        return self.head.pre.val

    def isEmpty(self) -> bool:
        return self.len==0

    def isFull(self) -> bool:
        return self.len==self.k


# Your MyCircularQueue object will be instantiated and called as such:
# obj = MyCircularQueue(k)
# param_1 = obj.enQueue(value)
# param_2 = obj.deQueue()
# param_3 = obj.Front()
# param_4 = obj.Rear()
# param_5 = obj.isEmpty()
# param_6 = obj.isFull()